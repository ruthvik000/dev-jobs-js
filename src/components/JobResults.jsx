import React from 'react';
import DateFormat from "./DateFormat";
import axios from "axios";
import JobInfo from "./JobInfo";

// Component to render search results.
function JobResults(props) {
    const [positionId, setPositionId] = React.useState('');
    const [error, setError] = React.useState(false);
    const [positionById, setPositionById] = React.useState(false);
    const [items, setItems] = React.useState([]);
    const [jobInfo, setJobInfo] = React.useState({});
    let content;

    React.useEffect(() => {
        axios.get(`https://cors-anywhere.herokuapp.com/https://jobs.github.com/positions.json?page=${props.pageNumber}${props.queryParam}`)
            .then(
                (result) => {
                    if (result.status !== 200) {
                        setError(true)
                    } else if (props.displayLoadMoreButton) {
                        if (result.data.length === 0) {
                            props.setDisplayLoadMoreButton(false);
                        }
                        setItems([...new Set([...items, ...result.data])])
                    } else {
                        setItems([...new Set(result.data)]);
                        props.setDisplayLoadMoreButton(true);
                    }
                },
            ).catch(() =>
            setError(true))
        // eslint-disable-next-line
    }, [props.pageNumber, props.queryParam])

    React.useEffect(() => {
        // Added this check is to not call the API when initially loading the application
        if (positionId !== '') {
            axios.get(`https://cors-anywhere.herokuapp.com/https://jobs.github.com/positions/${positionId}.json?markdown=true`)
                .then(
                    (result) => {
                        if (result.status !== 200) {
                            setError(true)
                        } else {
                            setJobInfo(result.data);
                            setPositionById(true);
                            setPositionId('')
                        }
                    },
                ).catch(() =>
                setError(true))
        }
    }, [positionId])

    function handleGridItemClick(id) {
        setPositionId(id)
        props.setJobSearchHeader(false);
    }

    function loadMoreItems() {
        props.setPageNumber(props.pageNumber + 1)
    }

    let errorMessage;
    if (error) {
        errorMessage =  <div>Results Failed to Load! Please Try Again..!</div>
    }

    if (positionById) {
        return <JobInfo jobInfo={jobInfo}/>
    } else {
        content = items.map((result, index) =>
            <div className='grid-content' key={index} onClick={() => handleGridItemClick(result.id)}>
                <img className='w3-round-small'
                     style={{
                         position: "absolute",
                         top: '-6%',
                         backgroundColor: "white",
                         left: '2%',
                         border: '1px solid #ececec',
                         padding: '5px'
                     }}
                     src={result.company_logo}
                     alt={result.company}
                     height="40px"/>
                <p className='days-type'>{DateFormat(new Date(result.created_at))} . {result.type}</p>
                <h4 className="job-title">{result.title}</h4>
                <p className='company-name'>{result.company}</p>
                <p className='job-location'>{result.location}</p>
            </div>
        );
    }

    let button;
    button = (items.length > 49 && props.displayLoadMoreButton) ?
        <button className="load-more-button" type="button" onClick={loadMoreItems}>Load More</button> : null

    return (
        <div style={{position: "relative"}}>
            <div className='grid'>
                {content}
                {errorMessage}
            </div>
            <div className="load-more">
                {button}
            </div>
        </div>
    );
}

export default (JobResults)