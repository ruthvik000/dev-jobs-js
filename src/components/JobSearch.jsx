import React from 'react';
import Modal from "./Modal";

// Component to render search results.
function JobSearch(props) {
    const [resize, setResize] = React.useState(window.innerWidth);
    const [input, setInput] = React.useState('');
    const [checked, setChecked] = React.useState(false);
    const [locationInput, setLocationInput] = React.useState('');
    const [isOpen, setIsOpen] = React.useState(false);
    const handleClick = () => setChecked(!checked)
    const prevQueryParam = usePreviousQueryParamValue(props.queryParam);


    const onresize = function (e) {
        return setResize(e.target.innerWidth);
    }

    function onChangeInput(e) {
        setInput(e.target.value)
    }

    function onChangeLocation(e) {
        if (e.target.value === 'Default Location') {
            getLocation()
        } else {
            props.setLatitude(0);
            props.setLongitude(0);
        }
        setLocationInput(e.target.value)
    }

    function getLocation() {
        if (navigator.geolocation) {
            navigator.geolocation.getCurrentPosition(showPosition);
        }
    }

    function showPosition(position) {
        props.setLatitude(position.coords.latitude);
        props.setLongitude(position.coords.longitude);
    }

    function handleSearchBar() {
        let string = '';
        if (checked) {
            string += `&full_time=${checked}`
        }
        if (input && input.length > 0) {
            string += `&search=${input}`
        }

        if (props.latitude && props.longitude) {
            string += `&lat=${props.latitude}&long=${props.longitude}`
        } else if (locationInput && locationInput.length > 0) {
            string += `&location=${locationInput}`
        }

        if ((prevQueryParam.length > 0 && string === '') || string.length > 0) {
            props.setDisplayLoadMoreButton(false);
        }
        props.setPageNumber(1);
        props.setQueryParam(string);
    }

    function usePreviousQueryParamValue(value) {
        const ref = React.useRef();
        React.useEffect(() => {
            ref.current = value;
        })
        return ref.current;
    }

    React.useEffect(() => {
        window.addEventListener("resize", onresize);
    }, [resize])

    const searchPlaceHolder = resize < 768 ? 'Filter by Title...' : 'Filter by Title, companies, expertise...';

    let jobSearchInputField =
        <span>
            <i className="fa fa-search job-search" aria-hidden="true"/>
        <input className='search-input'
               type="searchInput"
               id="searchName"
               name="searchName"
               onChange={onChangeInput}
               value={input}
               placeholder={searchPlaceHolder}/>
        </span>

    let remainingFields =
        <span>
            <i className="fa fa-map-marker" aria-hidden="true"/>
            <input className='search-location'
                   type="searchLocation"
                   id="locationName"
                   list='default'
                   name="locationName"
                   value={locationInput}
                   onChange={onChangeLocation}
                   placeholder="Filter by Location..."/>
                   <datalist id="default">
                       <option value="Default Location"/>
                   </datalist>
        </span>
    let content = null;
    if (props.jobSearchHeader) {
        if (resize < 768) {
            content = <form className="search-form">
                {jobSearchInputField}
                <button style={{border: "none"}}
                        className='search-button-filter'
                        onClick={() => setIsOpen(true)}
                        type="button">
                    <i style={{backgroundColor: "none"}} className="fa fa-filter" aria-hidden="true"/>
                </button>
                <Modal open={isOpen} onClose={() => setIsOpen(false)}>
                    <div>
                        <input className="checkbox-field"
                               type="checkbox"
                               id="fullTime"
                               name="fullTime"
                               checked={checked}
                               onChange={handleClick}
                        />
                        <label htmlFor="fullTime" className='checkbox-label'>Full Time Only</label>
                    </div>
                    {remainingFields}
                </Modal>
                <button className="btn search-button-mobile" type="button" onClick={handleSearchBar}>
                    <i className="fa fa-search search-icon-mobile"/>
                </button>
            </form>
        } else {
            content =
                <form className="search-form">
                    {jobSearchInputField}
                    {remainingFields}
                    <span>
                        <input className="checkbox-field"
                               style={{display: 'inline-block', margin: '5px'}}
                               type="checkbox"
                               id="fullTime"
                               name="fullTime"
                               checked={checked}
                               onChange={handleClick}
                        />
                        <label htmlFor="fullTime" className='checkbox-label'>Full Time Only</label>
                        <input className='search-button' type="button" value="Search" onClick={handleSearchBar}/>
                    </span>
                </form>
        }
    }
    return (content);
}

export default (JobSearch)