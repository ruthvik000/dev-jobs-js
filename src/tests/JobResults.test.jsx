import React from "react";
import {shallow} from 'enzyme'
import JobResults from "../components/JobResults";
import axios from 'axios';
import JobSearch from "../components/JobSearch";

const results = [
    {
        company_logo: 'xyz',
        company: 'Company 1',
        title: 'Java Developer',
        created_at: '2/29/2021 4:52:48 PM UTC',
        type: 'Full Time',
        location: 'New York'
    },
    {
        company_logo: 'xyza',
        company: 'Company 2',
        title: 'React Developer',
        created_at: '3/29/2021 4:52:48 PM UTC',
        type: 'Contract',
        location: 'Kansas'
    },
]

describe('Job Results component', () => {
    it('should match the snapshot with zero results', () => {
        jest.spyOn(React, 'useState')
            .mockReturnValueOnce(['', {}])
            .mockReturnValueOnce([false, {}])
            .mockReturnValueOnce([false, {}])
            .mockReturnValueOnce([[], {}])
            .mockReturnValueOnce([{}, {}])
            .mockReturnValueOnce([1, {}])
        axios.get = jest.fn(() => Promise.resolve({status: 200}, {data: []}))
        const container = shallow(<JobResults />);
        expect(container.html()).toMatchSnapshot();
    });

    it('should match the snapshot with results', () => {
        jest.spyOn(React, 'useState')
            .mockReturnValueOnce(['', {}])
            .mockReturnValueOnce([false, {}])
            .mockReturnValueOnce([false, {}])
            .mockReturnValueOnce([results, {}])
            .mockReturnValueOnce([{}, {}])
            .mockReturnValueOnce([1, {}])
        axios.get = jest.fn(() => Promise.resolve({status: 200}, {data: [results]}))
        const container = shallow(<JobResults />);
        expect(container.html()).toMatchSnapshot();
    });

    it('should match the snapshot with results and load more button', () => {
        jest.spyOn(React, 'useState')
            .mockReturnValueOnce(['', {}])
            .mockReturnValueOnce([false, {}])
            .mockReturnValueOnce([false, {}])
            .mockReturnValueOnce([results, {}])
            .mockReturnValueOnce([{}, {}])
            .mockReturnValueOnce([1, {}])
        axios.get = jest.fn(() => Promise.resolve({status: 200}, {data: [results]}))
        const container = shallow(<JobResults displayLoadMoreButton={true} />);
        expect(container.html()).toMatchSnapshot();
    });

    it('should match the snapshot with empty div for job info', () => {
        jest.spyOn(React, 'useState')
            .mockReturnValueOnce(['12345', {}])
            .mockReturnValueOnce([false, {}])
            .mockReturnValueOnce([true, {}])
            .mockReturnValueOnce([[], {}])
            .mockReturnValueOnce([results[0], {}])
            .mockReturnValueOnce([1, {}])
        axios.get = jest.fn(() => Promise.resolve({status: 200}, {data: [results]}))
        const container = shallow(<JobResults />);
        expect(container.html()).toMatchSnapshot();
    });

    describe('validate classes', () => {
        jest.spyOn(React, 'useState')
            .mockReturnValueOnce(['', {}])
            .mockReturnValueOnce([false, {}])
            .mockReturnValueOnce([false, {}])
            .mockReturnValueOnce([results, {}])
            .mockReturnValueOnce([{}, {}])
            .mockReturnValueOnce([1, {}])
        axios.get = jest.fn(() => Promise.resolve({status: 200}, {data: [results]}))
        const container = shallow(<JobResults />);

        it('should have an grid content class field', () => {
            expect(container.find('div[className="grid-content"]').length).toEqual(2);
        });

        it('should have an image class field', () => {
            expect(container.find('img[className="w3-round-small"]').length).toEqual(2);
        });

        it('should have an days and type class field', () => {
            expect(container.find('p[className="days-type"]').length).toEqual(2);
        });

        it('should have an company name class field', () => {
            expect(container.find('p[className="company-name"]').length).toEqual(2);
        });

        it('should have an job location class field', () => {
            expect(container.find('p[className="job-location"]').length).toEqual(2);
        });
    });
});