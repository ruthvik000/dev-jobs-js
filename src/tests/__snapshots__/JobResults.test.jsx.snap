// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Job Results component should match the snapshot with empty div for job info 1`] = `"<div><div class=\\"job-info\\"><img class=\\"w3-round-small job-info-image\\" style=\\"float:left\\" src=\\"xyz\\" alt=\\"Company 1\\" height=\\"100px\\" width=\\"100px\\"/><div class=\\"jobsite\\"><h2 class=\\"job-details-title\\">Company 1</h2><a target=\\"_blank\\" rel=\\"noopener noreferrer\\"><button class=\\"company-url-button\\">Company Site</button></a></div></div><div class=\\"job-description-body\\"><a target=\\"_blank\\" rel=\\"noopener noreferrer\\"><button class=\\"apply-now-button\\">Apply Now</button></a><p class=\\"days-type\\">1 month ago . Full Time</p><h3 class=\\"job-title\\">Java Developer</h3><p class=\\"job-location\\">New York</p><p class=\\"job-description\\"></p></div><div class=\\"how-to-apply\\"><h5>How to Apply</h5><p></p></div></div>"`;

exports[`Job Results component should match the snapshot with results 1`] = `"<div style=\\"position:relative\\"><div class=\\"grid\\"><div class=\\"grid-content\\"><img class=\\"w3-round-small\\" style=\\"position:absolute;top:-6%;background-color:white;left:2%;border:1px solid #ececec;padding:5px\\" src=\\"xyz\\" alt=\\"Company 1\\" height=\\"40px\\"/><p class=\\"days-type\\">1 month ago . Full Time</p><h4 class=\\"job-title\\">Java Developer</h4><p class=\\"company-name\\">Company 1</p><p class=\\"job-location\\">New York</p></div><div class=\\"grid-content\\"><img class=\\"w3-round-small\\" style=\\"position:absolute;top:-6%;background-color:white;left:2%;border:1px solid #ececec;padding:5px\\" src=\\"xyza\\" alt=\\"Company 2\\" height=\\"40px\\"/><p class=\\"days-type\\">2 weeks ago . Contract</p><h4 class=\\"job-title\\">React Developer</h4><p class=\\"company-name\\">Company 2</p><p class=\\"job-location\\">Kansas</p></div></div><div class=\\"load-more\\"></div></div>"`;

exports[`Job Results component should match the snapshot with results and load more button 1`] = `"<div style=\\"position:relative\\"><div class=\\"grid\\"><div class=\\"grid-content\\"><img class=\\"w3-round-small\\" style=\\"position:absolute;top:-6%;background-color:white;left:2%;border:1px solid #ececec;padding:5px\\" src=\\"xyz\\" alt=\\"Company 1\\" height=\\"40px\\"/><p class=\\"days-type\\">1 month ago . Full Time</p><h4 class=\\"job-title\\">Java Developer</h4><p class=\\"company-name\\">Company 1</p><p class=\\"job-location\\">New York</p></div><div class=\\"grid-content\\"><img class=\\"w3-round-small\\" style=\\"position:absolute;top:-6%;background-color:white;left:2%;border:1px solid #ececec;padding:5px\\" src=\\"xyza\\" alt=\\"Company 2\\" height=\\"40px\\"/><p class=\\"days-type\\">2 weeks ago . Contract</p><h4 class=\\"job-title\\">React Developer</h4><p class=\\"company-name\\">Company 2</p><p class=\\"job-location\\">Kansas</p></div></div><div class=\\"load-more\\"></div></div>"`;

exports[`Job Results component should match the snapshot with zero results 1`] = `"<div style=\\"position:relative\\"><div class=\\"grid\\"></div><div class=\\"load-more\\"></div></div>"`;
